### DISCUSSION TITLE
general comments for improvements

### DISCUSSION THREAD

**Discussion by orefalo** [lmorg/murex#705] [Ideas]
Hi Laurence, 

 I am sharing a few general comments prior to 5.0
  major releases are typically an opportunity to deprecate, I realize this discussion may come late.

- `signalTrap` should be renamed `onSignalTrap` to be consistent with the event handler naming convention
- Since Murex's special features seems to be under the `murex-*` umbrella, I would rename  `fid-kill`, `fid-killall`, `fid-list`
- casting: I know this is a non-documented feature: Everyone with recognize `(type)` for casting, rather than `:type:`
- There is a lot of different ways to do things

ie.

**semicolumns**

```
 murex-package: update
 murex-package update
```

**assignment**

```
V=34
V:=34
34 -> set V
```

**variable access**
```
ENV.MYVAR
$ENV.MYVAR
```
I personally believe `$ENV` is better and consistent

**export**
```
export V=45
45 -> export V
ENV.V = 45
$ENV.V = 45
```

The above are not issue.. Just concerned that these different syntaxes may create a breaking change in the future.

will stop here, we have plenty to discuss ;-)

**Comment by lmorg** [comment]
> semicolons
> ```
> murex-package: update
> murex-package update
> ```

Personally, I do already consider the semi-colon here deprecated. However I don't want to remove parser support for it because that would break backwards compatibility. So for now it will be an undocumented feature that will slowly bleed out of existence.

> assignment
> ```
> V=34
> V:=34
> 34 -> set V
There's really only two ways:
1. as an expression: `var = expression`
2. as a statement: `set var = parameters...`

The expression format supports both with and without `$` -- eventually I want to deprecate all usage without a dollar but that would require a lot of effort refining the edge cases with the parser (basically when to return a variable _name_ and when to return the variable _value_).

The statement variation supports running as a function or expression, but that's not really any different to `grep` supporting taking input from a parameter (well, file which you name as a parameter) or STDIN.

I do want to deprecate the function side of `set` so it's only really used as a method though. There isn't much benefit in `set var = ...` vs `var = ...`, however `command line -> command line -> set var` does have specific advantages (and disadvantages) compared with `[set] var = ...`

The `:=` operator does something different to `=`. The former preserves the data type and format if the variable exists, whereas the latter overwrites it.

eg
```
$ENV.PATH = %[ @ENV.PATH, /foo/bar ]
```
would break your session because you're converting $PATH to a JSON array. whereas
```
$ENV.PATH := %[ @ENV.PATH, /foo/bar ]
```
would work as expected because it's taking a JSON array, checking what type $PATH is (`paths`) and then converts the JSON array to a `paths` array.

Ideally this would be the default behaviour for `=` but a bunch of my automated tests failed when I did this, which made me concerned how significant a breaking change it might be.

To be honest, I'm not sold on `:=` as an operator. It's ridiculously easy to get confused with `=`. So I'll mark this feature as experimental in this release so that we can change in a later release (features's marked as experimental aren't subject to the same backwards compatibility assurances)

> variable access
> ```
> ENV.MYVAR
> $ENV.MYVAR
> ```

This is the parser problem I raised above. How do you know if `command $var` is after the variable name, or the variable contents? The parser wouldn't know. Traditional shells worked on the basis that assignments were without the dollar and retrieval included the dollar. Eventually I'll move everything over to the dollar but I need figure out how to fix all the edge cases first (I do have some ideas here already but they're a bit to major to make it into v5.0)

> I personally believe `$ENV` is better and consistent

I agree :)

> export
> ```
> export V=45
> 45 -> export V
> ENV.V = 45
> $ENV.V = 45
> ```

You can also set env's via `set`, eg `45 -> set $ENV.V` :)

I think `$ENV` should be the idiomatic way but we cannot deprecate `export` because I think that's handy for peoples muscle memory.

---

> signalTrap -> onSignalTrap

I was in two minds about whether to include `on` or not. The reason I didn't was because that was more an accident / coincidence that the others are all prefixed `on` rather than a convention I set out to do. But I think you're right that keeping things uniform would be nicer. Maybe `onTrappedSignal` because that reads slightly better, but it's the same format as you suggested.

> casting: I know this is a non-documented feature: Everyone with recognize `(type)` for casting, rather than `:type:`

I can't say I love `:type:` syntax either. There was a discussion about this one here: https://github.com/lmorg/murex/issues/484

**Reply by orefalo** [reply]
Imho: major versions are there to deprecate. Can be done in two steps - warning, then removal

**Reply by lmorg** [reply]
I agree. And this major version does come with potential breaking changes due to changes to the parser. But expressions and $ENV/$GLOBAL syntax are less than a year old. If I'm going to offer any kind of stability claims (and most alt shells don't so that's one major area Murex stands out) then I need to give these features a little more time to embed in before I even start pestering users with noisy warnings.

Perhaps your idea of a deprecations page is a good starting point? Then we can deprecate in 3 stages:
- notice of deprecation on the website (both the changelog and the deprecations page)
- warning of deprecation in the runtime (messages emitted by the parser)
- removal of feature

That could span maybe ~2 years but it gives plenty of time for even infrequent code to be updated?

For what it's worth, the Murex version constraint field I've added to modules is intended to help here too. In fact due to the way the parser is written, it's entirely possible for me to run multiple versions of the parser in parallel (ie if script is tagged for version `=4` then run the `4.x.x` parser rather than `latest`)

**Comment by lmorg** [comment]
I think I've cracked the `$var` problem:
```
» a [1..3] -> foreach --jmap $bob {out $bob}{out $bob}
{
    "1": "1",
    "2": "2",
    "3": "3"
}
```
(code in `develop`)

**Reply by orefalo** [reply]
Nice

**Reply by orefalo** [reply]
It works perfectly for my use cases/modules
